### PR TITLE
[jit] Add a utility class SourceRef to represent Source as keys

### PR DESCRIPTION
**Pull request by zhxchen17**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57398 [jit] Add Python API for ScriptProfile
* #57397 [jit] Implement ScriptProfile to collect instruction profiles.
* **#57396 [jit] Add a utility class SourceRef to represent Source as keys**

Summary:
A new type SourceRef is introduced to represent a unique identifier to source
text. The type holds refcount to underlying source, and supports comparators
and hash functions, such that it can be used in C++ and Python maps. In later
diffs we will use this to aggregate and print profiling information.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28133578](https://our.internmc.facebook.com/intern/diff/D28133578)

### CODE REVIEW

File: torch/csrc/jit/frontend/source_ref.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+
+#include <c10/macros/Export.h>
+#include <torch/csrc/jit/frontend/source_range.h>
+
```
Comment (nikithamalgifb) [inline torch/csrc/jit/frontend/source_ref.h:8]: Can you add a small comment on what this file does ?

File: torch/csrc/jit/frontend/source_ref.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  }
+  bool operator<(const Source& other) const {
+    return source_.get() < &other;
+  }
+  friend bool operator<(const Source& other, const SourceRef& self) {
+    return &other < self.source_.get();
+  }
+  bool operator<(const SourceRef& other) const {
+    return *this < *other.source_.get();
+  }
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/source_ref.h:27]: Curious why these comparators are needed.

**Review by gmagogsfm — APPROVED** [review]

File: torch/csrc/jit/frontend/source_ref.h (hunk 2 above)
Comment (zhxchen17) [reply torch/csrc/jit/frontend/source_ref.h:27]: I need to put this type as keys in std::map, and std::map is comparison based.

**Comment by zhxchen17** [comment]
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28133578).

**Comment by zhxchen17** [comment]
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28133578).

**Comment by zhxchen17** [comment]
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28133578).

**Comment by zhxchen17** [comment]
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28133578).

**Comment by zhxchen17** [comment]
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28133578).

**Pull request closed without merging**